Capi 23/fix/basic logging

galaxie opened the pull request:
Добавлено какое-никакое логирование и спеки.

galaxie committed:
CAPI-23 Add understandable messages in case of bad requests

galaxie committed:
CAPI-23 Add specs to shut up dialyzer and add some minor code refactoring

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/api.mustache:
```diff
@@ -17,6 +17,11 @@ request_params('{{operationId}}') ->
 request_params(_) ->
     error(unknown_operation).
 
+-spec request_param_info(OperationID :: atom(), Name :: atom()) -> #{
```
Не проще ли на этом уровне ввести типы `operation_id`, `parameter` и т.д., да ещё возможно и экспортнуть их.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/api.mustache:
```diff
@@ -17,6 +17,11 @@ request_params('{{operationId}}') ->
 
+-spec request_param_info(OperationID :: atom(), Name :: atom()) -> #{
+    source => qs_val | binding | header | body,
+    rules => [any()]
```
`any()` выглядит подозрительно.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/api.mustache:
```diff
@@ -50,6 +55,9 @@ request_param_info('{{operationId}}', {{^isBodyParam}}'{{baseName}}'{{/isBodyPar
     error({unknown_param, OperationID, Name}).
 
+-spec populate_request(OperationID :: atom(), Req :: cowboy_req:req(), ValidatorState :: jesse_state:state())
+-> {ok, Model :: #{}, Req :: cowboy_req:req()} | {error, Reason :: any(), Req :: cowboy_req:req()}.
```
Мне кажется лучше всегда в таких ситуациях ориентироваться на woody, практика с `{result_ok() | result_error(), context()}` выглядит логичнее.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/api.mustache:
```diff
@@ -117,8 +124,8 @@ validate( Rule = {type, 'integer'}, Name, Value, _ValidatorState) ->
     catch
-        _:Error ->
-            validation_error(Rule, {Name, Error})
+        _:_ ->
```
В таких ситуациях всегда есть риск поймать слишком много ситуаций, в том числе ложно отрицательных.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/api.mustache:
```diff
@@ -117,8 +124,8 @@ validate( Rule = {type, 'integer'}, Name, Value, _ValidatorState) ->
+            validation_error(Rule, Name)
     end;
 
 validate( Rule = {type, 'float'}, Name, Value, _ValidatorState) ->
```
А зачем здесь пробелы после первой скобки?

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/api.mustache:
```diff
@@ -220,19 +227,38 @@ validate( Rule = {pattern, Pattern}, Name, Value, _ValidatorState) ->
+-spec validation_error(Rule :: any(), Name :: any(), Info :: #{}) -> no_return().
+
+validation_error(ViolatedRule, Name, Info) ->
+    throw({wrong_param, Name, ViolatedRule, Info}).
```
...Тем более ты здесь явно бросаешь исключения определённого класса.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/default_logic_handler.mustache:
```diff
@@ -11,10 +11,16 @@
 {{/authMethods}}
 
+-spec handle_request(OperationID :: atom(), Req :: cowboy_req:req(), Context :: #{})
+-> {Code :: integer(), Headers :: [], Body :: binary()}.
```
Список чего?

keynslug replied:
Ну и к тому же, если ты в интерфейсе предполагаешь получение хэдеров от хэндлера и отдачу их наверх, ковбою, _в неизменном виде_, то здесь должны быть типы ковбоя.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/handler.mustache:
```diff
@@ -19,32 +19,49 @@
-    logic_handler,
-    validator_state,
-    context=#{}
+    operation_id :: atom(),
```
Очень напрашиваются общие типы.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/handler.mustache:
```diff
@@ -19,32 +19,49 @@
     {[], Req, State}.
 
+-spec is_authorized(Req :: cowboy_req:req(), State :: #state{})
+-> {Value :: any(), Req :: cowboy_req:req(), State :: #state{}}.
```
Сверхподохрительные спеки, у него же вот [здесь](https://github.com/ninenines/cowboy/blob/master/doc/src/manual/cowboy_rest.asciidoc) они не такие свободные.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/handler.mustache:
```diff
@@ -74,28 +95,69 @@ valid_content_headers(Req0, State = #state{operation_id = '{{operationId}}'}) ->
+    {error, Reason :: any()}.
+
+-spec process_response(Response :: handler_response(), cowboy_req:req(), #state{}) -> {halt, cowboy_req:req(), #state{}}.
+
```
Лишние строки какие-то.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/default_logic_handler.mustache:
```diff
@@ -11,10 +11,16 @@
 handle_request(OperationID, Req, Context) ->
-    io:format(user, "Got request to process: ~p~n", [{OperationID, Req, Context}]),
-    {501, [], <<"Not Implemented">>}.
+    Message = io_lib:format("Got not implemented request to process: ~p~n", [{OperationID, Req, Context}]),
```
А чем тебе форматирование средствами самого `error_logger` не угодило?

galaxie replied:
Строка длинная получается

keynslug replied:
Где именно она длинная получается?

galaxie replied:
~~Ой, все~~ Если одним выражением делать. Но твоя мысль понятна, я переделаю

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/handler.mustache:
```diff
@@ -74,28 +95,69 @@ valid_content_headers(Req0, State = #state{operation_id = '{{operationId}}'}) ->
+            Reason = io_lib:format("Unable to process request for ~p: ~p", [OperationID, Message]),
+            report(error, Reason),
+
+            {ok, Req} = cowboy_req:reply(400, Req0),
```
А мы не хотим говорить, что именно не так, или забили пока?

galaxie replied:
Я собираюсь забить, так как не представляю пока, как отдать тут честное сообщение об ошибке по спецификации

keynslug replied:
А как выглядит типичный `Message`?

galaxie committed:
CAPI-23
	Fix logging to use error_logger:msg functionality only.
	Refactor specs and missed spec.
	Add minor refactoring.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/handler.mustache:
```diff
@@ -111,11 +173,19 @@ handle_request_json(
                 {Code, Headers, Body} ->
```
А зачем тут кейс, если только один вариант? )

galaxie replied:
наследие

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/router.mustache:
```diff
@@ -2,6 +2,8 @@
 
 -export([get_paths/1]).
 
+-spec get_paths(LogicHandler :: atom()) ->  [{'_',[any()]},...].
```
Опять довольно общая спека. Зачем вообще такую писать, проще диализ отключить тогда. )

galaxie replied:
если бы я только мог )

galaxie replied:
вообще я такие херачил, когда не находил спек у `cowboy`

keynslug replied:
Но ты же сам знаешь, что ты возвращаешь.

galaxie committed:
CAPI-23 Fix codegen function name, add some specs

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/auth.mustache:
```diff
@@ -28,12 +33,18 @@ authorize_api_key(From, KeyParam, OperationID, Req0, LogicHandler) ->
 get_api_key(header, KeyParam, Req0) ->
-    {Headers, Req} = cowboy_req:headers(Req0),
-    {swagger_utils:get_opt({{packageName}}_utils:to_header(KeyParam), Headers), Req};
+  {Headers, Req} = cowboy_req:headers(Req0),
```
Отступы в два пробела зачем-то.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/auth.mustache:
```diff
@@ -28,12 +33,18 @@ authorize_api_key(From, KeyParam, OperationID, Req0, LogicHandler) ->
 get_api_key(qs_val, KeyParam, Req0) ->
-    {QS, Req} = cowboy_req:qs_vals(Req0),
-    {swagger_utils:get_opt(KeyParam, QS), Req}.
+  {QS, Req} = cowboy_req:qs_vals(Req0),
```
И тут.

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/auth.mustache:
```diff
@@ -3,22 +3,27 @@
     _ ->
-      case LogicHandler:authorize_api_key(ApiKey, OperationID) of
+      Result = {{packageName}}_logic_handler:authorize_api_key(
+        LogicHandler,
```
И тут два пробела.

keynslug commented:
В остальном, :shipit:

keynslug commented on modules/swagger-codegen/src/main/resources/erlang-server/api.mustache:
```diff
@@ -113,7 +114,7 @@ populate_request_param(OperationID, Name, Req0, ValidatorState) ->
 -spec validate_response(
     OperationID :: operation_id(),
-    Code :: integer(),
+    Code :: non_neg_integer(),
```
`200..599`?

galaxie committed:
CAPI-23 Add proper formatting in auth module. Limit http code range in specs

keynslug commented:
:shipit:
